src/Blocks/BlockPiston.cpp

```diff
@@ -86,28 +73,129 @@ bool cBlockPistonHandler::GetPlacementBlockTypeMeta(
 {
-	// Examine each of the 12 blocks ahead of the piston:
-	for (int ret = 0; ret < PISTON_MAX_PUSH_DISTANCE; ret++)
+	switch (a_PistonMeta & 0x07)
 	{
-		BLOCKTYPE currBlock;
-		NIBBLETYPE currMeta;
-		AddPistonDir(a_PistonX, a_PistonY, a_PistonZ, pistonmeta, 1);
-		a_World->GetBlockTypeMeta(a_PistonX, a_PistonY, a_PistonZ, currBlock, currMeta);
-		if (cBlockInfo::IsPistonBreakable(currBlock))
+		case 0: return Vector3i( 0, -1,  0);
+		case 1: return Vector3i( 0,  1,  0);
+		case 2: return Vector3i( 0,  0, -1);
+		case 3: return Vector3i( 0,  0,  1);
+		case 4: return Vector3i(-1,  0,  0);
+		case 5: return Vector3i( 1,  0,  0);
+		default:
 		{
-			// This block breaks when pushed, extend up to here
-			return ret;
+			LOGWARNING("%s: invalid direction %d, ignoring", __FUNCTION__, a_PistonMeta & 0x07);
+			return Vector3i();
 		}
-		if (!CanPush(currBlock, currMeta))
+	}
+}
+
+
+
+
+
+void cBlockPistonHandler::PushBlocks(const std::unordered_set<Vector3i, VectorHasher<int>> & a_BlocksToPush,
+	cWorld * a_World, const Vector3i & a_PushDir
+)
+{
+	// Sort blocks to move the blocks first, which are farest away from the piston
+	// This prevents the overwriting of existing blocks
+	std::vector<Vector3i> sortedBlocks(a_BlocksToPush.begin(), a_BlocksToPush.end());
+	std::sort(sortedBlocks.begin(), sortedBlocks.end(), [a_PushDir](const Vector3i & a, const Vector3i & b)
+	{
+		return a.Dot(a_PushDir) > b.Dot(a_PushDir);
+	});
+
+	// Move every block
+	BLOCKTYPE moveBlock;
+	NIBBLETYPE moveMeta;
+	for (Vector3i & moveBlockPos : sortedBlocks)
+	{
+		a_World->GetBlockTypeMeta(moveBlockPos.x, moveBlockPos.y, moveBlockPos.z, moveBlock, moveMeta);
+		a_World->SetBlock(moveBlockPos.x, moveBlockPos.y, moveBlockPos.z, E_BLOCK_AIR, 0);
+
+		moveBlockPos += a_PushDir;
+		if (cBlockInfo::IsPistonBreakable(moveBlock))
+		{
+			// Block is breakable, drop it
+			cBlockHandler * Handler = BlockHandler(moveBlock);
+			if (Handler->DoesDropOnUnsuitable())
+			{
+				cChunkInterface ChunkInterface(a_World->GetChunkMap());
+				cBlockInServerPluginInterface PluginInterface(*a_World);
+				Handler->DropBlock(ChunkInterface, *a_World, PluginInterface, nullptr,
+					moveBlockPos.x, moveBlockPos.y, moveBlockPos.z
+				);
+			}
+		} else
 		{
-			// This block cannot be pushed at all, the piston can't extend
-			return -1;
+			// Not breakable, just move it
+			a_World->SetBlock(moveBlockPos.x, moveBlockPos.y, moveBlockPos.z, moveBlock, moveMeta);
 		}
 	}
-	// There is no space for the blocks to move, piston can't extend
-	return -1;
+}
+
+
+
+
+
+bool cBlockPistonHandler::CanPushBlock(
+	int a_BlockX, int a_BlockY, int a_BlockZ, cWorld * a_World, bool a_RequirePushable,
+	std::unordered_set<Vector3i, VectorHasher<int>> & a_BlocksPushed, const Vector3i & a_PushDir
+)
+{
+	const static std::array<Vector3i, 6> pushingDirs = {{ Vector3i(-1, 0, 0), Vector3i(1, 0, 0), Vector3i(0, -1, 0), Vector3i(0, 1, 0),
+		Vector3i(0, 0, -1), Vector3i(0, 0, 1) }};
```

Comment:
This would look like this:

    ... = 
    {
    {
        ....
    }}

Or similar to this(I could indent the inner braces). But the check code doesn't allow to have two opening braces in the same line.